File: clients/go/connecto/hub.go
package connecto

import (
	"encoding/json"
	"fmt"
	"net/url"
	"strings"
	"sync"

	"github.com/gorilla/websocket"
	"github.com/google/uuid"
)

const recordSeparator = '\x1e'

type invocationState struct {
	result chan json.RawMessage
	err    chan error
}

type HubConnection struct {
	conn        *websocket.Conn
	mu          sync.Mutex
	invocations map[string]*invocationState
	broadcasts  map[string]chan json.RawMessage
	done        chan struct{}
}

func NewHubConnection(baseURL, path, token string) (*HubConnection, error) {
	wsURL := strings.ReplaceAll(baseURL, "http://", "ws://")
	wsURL = strings.ReplaceAll(wsURL, "https://", "wss://")

	u, err := url.Parse(wsURL + path + "?access_token=" + token)
	if err != nil {
		return nil, fmt.Errorf("parse url: %w", err)
	}

	conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("dial: %w", err)
	}

	// Send handshake
	handshake, _ := json.Marshal(handshakeRequest{Protocol: "json", Version: 1})
	err = conn.WriteMessage(websocket.TextMessage, append(handshake, byte(recordSeparator)))
	if err != nil {
		return nil, fmt.Errorf("handshake write: %w", err)
	}

	hc := &HubConnection{
		conn:        conn,
		invocations: make(map[string]*invocationState),
		broadcasts:  make(map[string]chan json.RawMessage),
		done:        make(chan struct{}),
	}

	go hc.readLoop()
	return hc, nil
}

func (h *HubConnection) Invoke(target string, args ...interface{}) (json.RawMessage, error) {
	invID := uuid.New().String()
	resultCh := make(chan json.RawMessage, 1)
	errCh := make(chan error, 1)

	h.mu.Lock()
	h.invocations[invID] = &invocationState{result: resultCh, err: errCh}
	h.mu.Unlock()

	msg := invocationMessage{
		Type:         1,
		InvocationID: invID,
		Target:       target,
		Arguments:    args,
	}
	data, _ := json.Marshal(msg)
	frame := append(data, byte(recordSeparator))

	h.mu.Lock()
	err := h.conn.WriteMessage(websocket.TextMessage, frame)
	h.mu.Unlock()
	if err != nil {
		return nil, fmt.Errorf("invoke write: %w", err)
	}

	select {
	case result := <-resultCh:
		return result, nil
	case err := <-errCh:
		return nil, err
	}
}

func (h *HubConnection) WaitForBroadcast(event string) (json.RawMessage, error) {
	h.mu.Lock()
	if _, ok := h.broadcasts[event]; !ok {
		h.broadcasts[event] = make(chan json.RawMessage, 100)
	}
	ch := h.broadcasts[event]
	h.mu.Unlock()

	select {
	case val := <-ch:
		return val, nil
	case <-h.done:
		return nil, fmt.Errorf("connection closed")
	}
}

func (h *HubConnection) Close() {
	close(h.done)
	h.conn.Close()
}

func (h *HubConnection) readLoop() {
	defer h.Close()
	for {
		_, message, err := h.conn.ReadMessage()
		if err != nil {
			return
		}

		frames := strings.Split(string(message), string(recordSeparator))
		for _, frame := range frames {
			if frame == "" || frame == "{}" {
				continue
			}

			var msg completionMessage
			if err := json.Unmarshal([]byte(frame), &msg); err != nil {
				continue
			}

			switch msg.Type {
			case 1:
				// Server-to-client invocation (broadcast event)
				if msg.Target != nil && msg.Arguments != nil {
					h.mu.Lock()
					ch, ok := h.broadcasts[*msg.Target]
					h.mu.Unlock()
					if ok {
						var val interface{}
						if len(msg.Arguments) == 1 {
							val = msg.Arguments[0]
						} else {
							val = msg.Arguments
						}
						data, _ := json.Marshal(val)
						select {
						case ch <- data:
						default:
						}
					}
				}
			case 3:
				// Completion
				if msg.InvocationID != nil {
					h.mu.Lock()
					state, ok := h.invocations[*msg.InvocationID]
					if ok {
						delete(h.invocations, *msg.InvocationID)
					}
					h.mu.Unlock()

					if ok && state != nil {
						if msg.Error != nil {
							state.err <- fmt.Errorf(*msg.Error)
						} else {
							var result json.RawMessage
							if msg.Result != nil {
								result, _ = json.Marshal(msg.Result)
							}
							state.result <- result
						}
					}
				}
			}
		}
	}
}

File: clients/rust/connecto-client/src/hub.rs
use std::collections::HashMap;
use std::sync::Arc;

use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tokio::sync::{mpsc, oneshot, Mutex};
use tokio_tungstenite::{connect_async, tungstenite::Message, MaybeTlsStream, WebSocketStream};
use tokio::net::TcpStream;
use url::Url;

const RECORD_SEPARATOR: char = '\x1e';

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct HandshakeRequest {
    protocol: String,
    version: u32,
}

#[derive(Deserialize)]
struct CompletionFrame {
    #[serde(rename = "type")]
    msg_type: u32,
    invocation_id: Option<String>,
    result: Option<Value>,
    error: Option<String>,
    target: Option<String>,
    arguments: Option<Vec<Value>>,
}

pub type WsWrite = futures_util::stream::SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>;
pub type WsRead = futures_util::stream::SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>;

pub struct HubConnection {
    pub write_tx: mpsc::UnboundedSender<String>,
    pub broadcast_tx: mpsc::UnboundedSender<(String, Value)>,
    invocations: Arc<Mutex<HashMap<String, oneshot::Sender<Result<Value, String>>>>>,
    broadcast_rx: Arc<Mutex<mpsc::UnboundedReceiver<(String, Value)>>>,
}

impl HubConnection {
    pub async fn connect(base_url: &str, path: &str, token: &str) -> Result<Self, String> {
        let ws_url = base_url
            .replace("http://", "ws://")
            .replace("https://", "wss://");
        let url = Url::parse(&format!("{}{}?access_token={}", ws_url, path, token))
            .map_err(|e| e.to_string())?;

        let (ws, _) = connect_async(url).await.map_err(|e| e.to_string())?;
        let (mut write, read) = ws.split();

        // Send handshake
        let handshake = serde_json::to_string(&HandshakeRequest {
            protocol: "json".to_string(),
            version: 1,
        })
        .unwrap();
        write
            .send(Message::Text(format!("{}{}", handshake, RECORD_SEPARATOR).into()))
            .await
            .map_err(|e| e.to_string())?;

        let invocations = Arc::new(Mutex::new(HashMap::new()));
        let (write_tx, mut write_rx) = mpsc::unbounded_channel::<String>();
        let (broadcast_tx, broadcast_rx) = mpsc::unbounded_channel::<(String, Value)>();

        // Spawn writer task
        let write_handle = tokio::spawn(async move {
            while let Some(msg) = write_rx.recv().await {
                let _ = write.send(Message::Text(msg.into())).await;
            }
        });

        let broadcast_rx = Arc::new(Mutex::new(broadcast_rx));

        // Spawn reader task
        let inv_clone = Arc::clone(&invocations);
        let broadcast_tx_clone = broadcast_tx.clone();
        tokio::spawn(async move {
            let mut read = read;
            while let Some(Ok(Message::Text(text))) = read.next().await {
                for frame in text.split(RECORD_SEPARATOR) {
                    if frame.is_empty() || frame == "{}" {
                        // handshake ack or empty
                        continue;
                    }
                    if let Ok(msg) = serde_json::from_str::<CompletionFrame>(frame) {
                        match msg.msg_type {
                            1 => {
                                // Invocation from server -> broadcast event
                                if let (Some(target), Some(args)) = (msg.target, msg.arguments) {
                                    let val = if args.len() == 1 {
                                        args.into_iter().next().unwrap()
                                    } else {
                                        Value::Array(args)
                                    };
                                    let _ = broadcast_tx_clone.send((target, val));
                                }
                            }
                            3 => {
                                // Completion
                                if let Some(inv_id) = msg.invocation_id {
                                    let mut map = inv_clone.lock().await;
                                    if let Some(tx) = map.remove(&inv_id) {
                                        match msg.error {
                                            Some(err) => { let _ = tx.send(Err(err)); }
                                            None => { let _ = tx.send(Ok(msg.result.unwrap_or(Value::Null))); }
                                        }
                                    }
                                }
                            }
                            _ => {}
                        }
                    }
                }
            }
        });

        Ok(Self {
            write_tx,
            broadcast_tx,
            invocations,
            broadcast_rx: Arc::clone(&broadcast_rx),
        })
    }

    pub async fn invoke(&self, target: &str, args: Vec<Value>) -> Result<Value, String> {
        let inv_id = uuid::Uuid::new_v4().to_string();
        let (tx, rx) = oneshot::channel();

        {
            let mut map = self.invocations.lock().await;
            map.insert(inv_id.clone(), tx);
        }

        let msg = serde_json::json!({
            "type": 1,
            "invocationId": inv_id,
            "target": target,
            "arguments": args,
        });
        let frame = format!("{}{}", serde_json::to_string(&msg).unwrap(), RECORD_SEPARATOR);
        self.write_tx.send(frame).map_err(|e| e.to_string())?;

        rx.await.map_err(|_| "Invocation cancelled".to_string())?
    }

    pub async fn wait_for_broadcast(&self, event: &str) -> Result<Value, String> {
        let mut rx = self.broadcast_rx.lock().await;
        loop {
            let (evt, val) = rx.recv().await.ok_or("Connection closed")?;
            if evt == event {
                return Ok(val);
            }
        }
    }
}
